habitat-lab/habitat/datasets/rearrange/samplers/receptacle.py

```diff
@@ -505,33 +505,161 @@ def get_all_scenedataset_receptacles(
     return receptacles
 
 
-def import_tri_mesh_ply(ply_file: str) -> TriangleMeshData:
+def import_tri_mesh(mesh_file: str) -> List[TriangleMeshData]:
     """
-    Returns a Tuple of (verts,indices) from a ply mesh using magnum trade importer.
+    Returns a list of Tuples of (verts,indices) from a mesh asset using magnum trade importer.
 
-    :param ply_file: The input PLY mesh file. NOTE: must contain only triangles.
+    :param mesh_file: The input meshes file. NOTE: must contain only triangles.
     """
     manager = mn.trade.ImporterManager()
     importer = manager.load_and_instantiate("AnySceneImporter")
-    importer.open_file(ply_file)
+    importer.open_file(mesh_file)
 
-    # TODO: We don't support mesh merging or multi-mesh parsing currently
-    if importer.mesh_count > 1:
-        raise NotImplementedError(
-            "Importing multi-mesh receptacles (mesh merging or multi-mesh parsing) is not supported."
+    mesh_data: List[TriangleMeshData] = []
+
+    def add_mesh_data(mesh_ix: int, abs_trans: Optional[mn.Matrix4] = None):
+        """
+        Load and transform mesh data into global data-structures.
+        """
+        mesh = importer.mesh(mesh_ix)
+
+        assert (
+            mesh.primitive == mn.MeshPrimitive.TRIANGLES
+        ), "Must be a triangle mesh."
+
+        transformed_positions = mesh.attribute(mn.trade.MeshAttribute.POSITION)
+        # apply the absolute transforms to the mesh data
+        if abs_trans is not None:
+            transformed_positions = [
+                abs_trans.transform_point(pos)
+                for pos in mesh.attribute(mn.trade.MeshAttribute.POSITION)
+            ]
+
+        mesh_data.append(
+            TriangleMeshData(
+                transformed_positions,
+                mesh.indices,
+            )
         )
 
-    mesh_ix = 0
-    mesh = importer.mesh(mesh_ix)
-    assert (
-        mesh.primitive == mn.MeshPrimitive.TRIANGLES
-    ), "Must be a triangle mesh."
+    # Import mesh data from a scene format (e.g. glTF)
+    scene_ix = importer.default_scene
+    if scene_ix != -1:
+        scene_data = importer.scene(scene_ix)
```

Comment:
With https://github.com/facebookresearch/habitat-sim/pull/2066 this whole code could be replaced with something like this, assuming the goal is to get back a list of meshes that have their positions transformed according to the scene hierarchy. Basically it's about porting [this example snippet](https://doc.magnum.graphics/magnum/namespaceMagnum_1_1SceneTools.html#ab1505270c538ebe54e23c78676d7a780) to Python:

```py
meshes = [importer.mesh(i) for i in range(importer.mesh_count)]

# Mesh referenced by mesh_assignments[i] has a corresponding transform in
# mesh_transformations[i]. Association to a particular node ID is stored in 
# scene.mapping(mn.trade.SceneField.MESH)[i], but it's not needed for anything
# here.
mesh_assignments: cr.containers.StridedArrayView1D = \
    scene.field(mn.trade.SceneField.MESH)
mesh_transformations: List[mn.Matrix4] = \
    mn.scenetools.flatten_transformation_hierarchy3d(scene, mn.trade.SceneField.MESH)
assert len(mesh_assignments) == len(mesh_transformations)

# A mesh can be referenced by multiple nodes, so this can't operate in-place.
flattened_meshes: List[mn.trade.MeshData] = []
for mesh_id, transformation in zip(mesh_assignments, mesh_transformations):
    flattened_meshes += [mn.meshtools.transform3d(mesh_id, transformation)]
```

The `flattened_meshes` is then very close to the `mesh_data` array here. You may however still want to store the `MeshData` instances or (references to) the position arrays from them instead of a `List[mn.Vector3]`, as that's *significantly* more memory-efficient (as far as I can see, each `mn.Vector3` instance stored in a list is 64 *bytes*, a lot more than the three floats per position when stored inside `MeshData`). 

If you'd want to reduce the memory use even more, you could discard also all unused attributes except positions. I.e. do this in the mesh loading code:

```py
mesh = ...
# This operation affects only the metadata -- the remaining attributes stay in 
# the vertex buffer but are no longer referenced and thus treated like padding
mesh = mn.meshtools.filter_only_attributes(mesh, [mn.trade.MeshAttribute.POSITION])
# This forces a repack of the mesh, throwing away all unused data (by default 
# it keeps the original layout including all padding if it's already interleaved)
mesh = mn.meshtools.interleave(mesh, mn.meshtools.InterleaveFlag.NONE)
```

(Wrote this off the top of my head, can't guarantee it compiles or works. But should be pretty close to a working state.)

Reply:
Thanks for the API updates and suggestions here.
I refactored the code to use the new APIs and took your suggestion to store the `MeshData` directly instead of lists.

However, I ran into some issues with the memory improvement suggestions. I think these could be added in a later PR, but for reference:
1. `mesh = mn.meshtools.filter_only_attributes(mesh, [mn.trade.MeshAttribute.POSITION])`
Seems to corrupt the `indices`:
```
len(self.mesh_data.indices) = 6
len(self.mesh_data.attribute(mn.trade.MeshAttribute.POSITION)) = 4
index = 0
self.mesh_data.indices[index] = 32349
```
2. `mesh = mn.meshtools.interleave(mesh, mn.meshtools.InterleaveFlag.NONE)` maybe need some additional plugin or compile flags?
```
AttributeError: module '_magnum.meshtools' has no attribute 'InterleaveFlag'
```

Reply:
Also, I'm getting the following terminal output even with `export MAGNUM_LOG=quiet`:

```
PluginManager::Manager: duplicate static plugin AssimpImporter, ignoring
PluginManager::Manager: duplicate static plugin AnySceneImporter, ignoring
PluginManager::Manager: duplicate static plugin AnyImageImporter, ignoring
Trade::GltfImporter::scene(): node 0 extras modelNormal property is Utility::JsonToken::Type::Object, skipping
Trade::GltfImporter::scene(): node 0 extras modelObb property is Utility::JsonToken::Type::Object, skipping
Trade::GltfImporter::scene(): node 1 extras modelNormal property is Utility::JsonToken::Type::Object, skipping
Trade::GltfImporter::scene(): node 1 extras modelObb property is Utility::JsonToken::Type::Object, skipping
Trade::GltfImporter::scene(): node 2 extras modelNormal property is Utility::JsonToken::Type::Object, skipping
Trade::GltfImporter::scene(): node 2 extras modelObb property is Utility::JsonToken::Type::Object, skipping

```

Reply:
> `export MAGNUM_LOG=quiet`

Unlike with `HABITAT_LOG`, this affects only the startup log when creating a GL context, not everything. Implementing that would either mean `getenv()` on every print (which is very slow), or need some global "logging context" singleton which -- while completely fine in Habitat -- is a rather nasty thing to have in a library / middleware setting and would introduce more problems than it would solve.

From the application perspective, silencing the output is possible with scoped redirection, such as this in C++ ([relevant docs](https://doc.magnum.graphics/corrade/classCorrade_1_1Utility_1_1Debug.html#Utility-Debug-scoped-output)):

```cpp
{
    Mn::Warning redirectWarnings{nullptr};
    // all warnings coming from Magnum are silenced now
}

// all warnings coming from Magnum are printed again
```

Hm, so maybe I could expose something similar to Python as well. I'm thinking it could be doable with a context manager and a `with` statement, although I'd need to look up how to do that with pybind. Would this make sense?

```py
import magnum as mn

with mn.redirect_warnings(None):
    # all warnings coming from Magnum are silenced now
    
# all warnings coming from Magnum are printed again
```

Another idea I had was to introduce `Flag::Quiet` on the importer interface, and make the plugins respect it. There's `Flag::Verbose` already. I'll add that, this seems good to have in any case.

> `Trade::GltfImporter::scene(): node 0 extras modelNormal property is Utility::JsonToken::Type::Object, skipping`

I was asking about these on Slack a while ago: https://cvmlp.slack.com/archives/C048W359BFC/p1677178392534799?thread_ts=1677014872.077569&cid=C048W359BFC

My goal is to make them all easily available through Python, so you could for example get them as `scene.field(importer.scene_field_for_name('modelNormal'))`, which also means having them implicitly preserved though all possible scene transformations and exported back when you save a file again. There's two options as I said in that thread -- either those get changed to not be in nested objects so the existing importer code can fetch them, or I update the glTF import/export code to handle them, somehow. I didn't get to the latter yet.

Reply:
> `module '_magnum.meshtools' has no attribute 'InterleaveFlag'`

It's `InterleaveFlags`, sorry. (I just typed that code snippet out from memory.)

> Seems to corrupt the `indices`

Uh oh. Working on getting that fixed :wink:  The root cause is that `meshtools.filter_*_attributes()` modify the metadata but reference the original mesh, and Python wasn't made aware of the original still being referenced. I'm making some additional changes to have these accidents caught automatically to prevent this from happening again.

I think this should work for now, I'll ping you once I have the overal robustness improved:

```py
mesh = ...
# `mesh_filtered` now refers to data in `mesh`
# (`mesh_filtered.vertex_data_flags` isn't `OWNED`)
# until i make a proper fix, it's important to use a different variable
# which should tell Python's GC to not delete the original
mesh_filtered = mn.meshtools.filter_only_attributes(mesh, [mn.trade.MeshAttribute.POSITION])

# makes the mesh standalone / self-owned again
# (`mesh.vertex_data_flags` is `OWNED`)
mesh = mn.meshtools.interleave(mesh_filtered, mn.meshtools.InterleaveFlags.NONE)
```

Reply:
Thanks, this appears to be working now.

Reply:
> Hm, so maybe I could expose something similar to Python as well. ...
> ```python
> with mn.redirect_warnings(None):
>     # all warnings coming from Magnum are silenced now
> ```
> ...
> Another idea I had was to introduce `Flag::Quiet` on the importer interface, and make the plugins respect it. ...

Either of these should functionally solve the problem from my perspective. In general a solution to quieting logs would be useful once we get to scaled training or running batched processes. :+1: 

> > `Trade::GltfImporter::scene(): node 0 extras modelNormal property is Utility::JsonToken::Type::Object, skipping`
> 
> I was asking about these on Slack a while ago: https://cvmlp.slack.com/archives/C048W359BFC/p1677178392534799?thread_ts=1677014872.077569&cid=C048W359BFC

I generally agree that accessing these nested objects would be valuable in the long run, but currently I just want to suppress the noise. :slightly_smiling_face: